File: LAB 6/Python/LINE_CLIPPING_MidpointSubdivision.py
import matplotlib.pyplot as plt
import numpy as np

XMIN, YMIN, XMAX, YMAX = 3, 3, 8, 8
INSIDE, LEFT, RIGHT, BOTTOM, TOP = 0, 1, 2, 4, 8
EPSILON = 0.01

LINES = [((2, 2), (10, 10)), ((1, 5), (10, 5)), ((5, 1), (5, 10)), ((9, 2), (9, 9)), ((2, 9), (9, 2))]

def compute_outcode(X, Y):
    CODE = INSIDE
    if X < XMIN: CODE |= LEFT
    elif X > XMAX: CODE |= RIGHT
    if Y < YMIN: CODE |= BOTTOM
    elif Y > YMAX: CODE |= TOP
    return CODE

def midpoint_subdivision(X1, Y1, X2, Y2):
    OUTCODE1 = compute_outcode(X1, Y1)
    OUTCODE2 = compute_outcode(X2, Y2)
    
    if not (OUTCODE1 | OUTCODE2):
        return [(X1, Y1, X2, Y2)]
    
    if OUTCODE1 & OUTCODE2:
        return []
    
    if abs(X2 - X1) < EPSILON and abs(Y2 - Y1) < EPSILON:
        return []
    
    XM = (X1 + X2) / 2
    YM = (Y1 + Y2) / 2
    
    SEGMENTS = []
    SEGMENTS.extend(midpoint_subdivision(X1, Y1, XM, YM))
    SEGMENTS.extend(midpoint_subdivision(XM, YM, X2, Y2))
    
    return SEGMENTS

plt.figure(figsize=(10, 8))
plt.grid(True)
plt.title('Midpoint Subdivision Line Clipping')
plt.xlabel('X-axis'); plt.ylabel('Y-axis')

plt.plot([XMIN, XMAX, XMAX, XMIN, XMIN], [YMIN, YMIN, YMAX, YMAX, YMIN], 'b-', linewidth=2, label='Clip Window')

for I, ((X1, Y1), (X2, Y2)) in enumerate(LINES):
    plt.plot([X1, X2], [Y1, Y2], 'r--', linewidth=1.5, alpha=0.5)
    CLIPPED_SEGMENTS = midpoint_subdivision(X1, Y1, X2, Y2)
    for SEG in CLIPPED_SEGMENTS:
        plt.plot([SEG[0], SEG[2]], [SEG[1], SEG[3]], 'g-', linewidth=2.5)

plt.xlim(0, 12); plt.ylim(0, 12)
plt.legend(['Clip Window', 'Original Lines', 'Clipped Lines'])
plt.show()

File: LAB 6/MATLAB/LINE_CLIPPING_Midpoint_Subdivision.m
clc; clear; close all;

XMIN = 3; YMIN = 3; XMAX = 8; YMAX = 8;
INSIDE = 0; LEFT = 1; RIGHT = 2; BOTTOM = 4; TOP = 8;
EPSILON = 0.01;

LINES = {[2, 2, 10, 10]; [1, 5, 10, 5]; [5, 1, 5, 10]; [9, 2, 9, 9]; [2, 9, 9, 2]};

function CODE = compute_outcode(X, Y, XMIN, YMIN, XMAX, YMAX, INSIDE, LEFT, RIGHT, BOTTOM, TOP)
    CODE = INSIDE;
    if X < XMIN, CODE = bitor(CODE, LEFT); end
    if X > XMAX, CODE = bitor(CODE, RIGHT); end
    if Y < YMIN, CODE = bitor(CODE, BOTTOM); end
    if Y > YMAX, CODE = bitor(CODE, TOP); end
end

function SEGMENTS = midpoint_subdivision(X1, Y1, X2, Y2, XMIN, YMIN, XMAX, YMAX, INSIDE, LEFT, RIGHT, BOTTOM, TOP, EPSILON)
    SEGMENTS = [];
    OUTCODE1 = compute_outcode(X1, Y1, XMIN, YMIN, XMAX, YMAX, INSIDE, LEFT, RIGHT, BOTTOM, TOP);
    OUTCODE2 = compute_outcode(X2, Y2, XMIN, YMIN, XMAX, YMAX, INSIDE, LEFT, RIGHT, BOTTOM, TOP);
    
    if ~bitor(OUTCODE1, OUTCODE2)
        SEGMENTS = [SEGMENTS; X1, Y1, X2, Y2];
        return;
    end
    
    if bitand(OUTCODE1, OUTCODE2)
        return;
    end
    
    if abs(X2 - X1) < EPSILON && abs(Y2 - Y1) < EPSILON
        return;
    end
    
    XM = (X1 + X2) / 2;
    YM = (Y1 + Y2) / 2;
    
    SEG1 = midpoint_subdivision(X1, Y1, XM, YM, XMIN, YMIN, XMAX, YMAX, INSIDE, LEFT, RIGHT, BOTTOM, TOP, EPSILON);
    SEG2 = midpoint_subdivision(XM, YM, X2, Y2, XMIN, YMIN, XMAX, YMAX, INSIDE, LEFT, RIGHT, BOTTOM, TOP, EPSILON);
    
    SEGMENTS = [SEGMENTS; SEG1; SEG2];
end

figure;
hold on; grid on; axis equal;
title('Midpoint Subdivision Line Clipping');
xlabel('X-axis'); ylabel('Y-axis');

plot([XMIN, XMAX, XMAX, XMIN, XMIN], [YMIN, YMIN, YMAX, YMAX, YMIN], 'b-', 'LineWidth', 2, 'DisplayName', 'Clip Window');

for I = 1:length(LINES)
    LINE = LINES{I};
    X1 = LINE(1); Y1 = LINE(2); X2 = LINE(3); Y2 = LINE(4);
    
    plot([X1, X2], [Y1, Y2], 'r--', 'LineWidth', 1.5);
    
    CLIPPED_SEGMENTS = midpoint_subdivision(X1, Y1, X2, Y2, XMIN, YMIN, XMAX, YMAX, INSIDE, LEFT, RIGHT, BOTTOM, TOP, EPSILON);
    for J = 1:size(CLIPPED_SEGMENTS, 1)
        SEG = CLIPPED_SEGMENTS(J, :);
        plot([SEG(1), SEG(3)], [SEG(2), SEG(4)], 'g-', 'LineWidth', 2.5);
    end
end

xlim([0, 12]); ylim([0, 12]);
legend('Clip Window', 'Location', 'best');
